### PR TITLE
fix(ctb): FDG Local Keys

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adjusts the `FaultDisputeGame`'s local key identifier `4` to post the disputed L2 output block number. The `op-program` consumes this local key to determine which block the derivation pipeline should [target](https://github.com/ethereum-optimism/optimism/blob/develop/op-program/client/driver/driver.go#L50), and ultimately [verifies](https://github.com/ethereum-optimism/optimism/blob/develop/op-program/client/driver/driver.go#L85-L95) the L2 claim against the L2 output submitted at this block number.

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* **#9670** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9670?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9670?utm_source=stack-comment).

**Review by Inphi — APPROVED** [review]
Thanks clabby!

It seems this code path never got hit during e2e. It would be nice to rely on solidity coverage reports for cannon e2e tests so similar problems don't occur in the future. Though I'm not even sure how we could begin to add this. Probably would require a non-trivial amount of changes to geth.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol [hunk 1]
```diff
@@ -325,15 +325,15 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
...
 
-            // If the starting position is 0 (invalid), the starting output root is genesis. Otherwise,
-            // we add the index at depth + 1 to the genesis block number to get the L2 block number.
+            // If the starting position is 0 (invalid), the disputed output root is at genesis + 1. Otherwise,
+            // we add the index at depth + 2 to the genesis block number to get the disputed L2 block number.
             uint256 l2Number = startingPos.raw() == 0
-                ? GENESIS_BLOCK_NUMBER
-                : GENESIS_BLOCK_NUMBER + startingPos.traceIndex(SPLIT_DEPTH) + 1;
+                ? GENESIS_BLOCK_NUMBER + 1
+                : GENESIS_BLOCK_NUMBER + startingPos.traceIndex(SPLIT_DEPTH) + 2;
```
Comment (ajsutton) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: It may make more sense to use `disputedPos + 1` instead of `startingPos + 2` here.

Comment (ajsutton) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: Can probably avoid the need for the special case genesis check too.

Comment (ajsutton) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: Actually this is definitely incorrect, it should just be:
```
uint256 l2Number = GENESIS_BLOCK_NUMBER + disputedPos.traceIndex(SPLIT_DEPTH) + 1;
```

A trace index of 0 is valid - it's the first block after genesis. The absolute prestate is effectively trace index -1.

**Comment by ajsutton** [comment]
I've added these changes into https://github.com/ethereum-optimism/optimism/pull/9687 with the fix to use the disputed block number and no exception for genesis.  That makes the new test that checks the preimages match start passing.

**Pull request merged**